---
title: "Building regulons"
author:
- name: Shang-Yang Chen
  email: chens179@gene.com
- name: Xiaosai Yao
  email: yaox19@gene.com
date: "February 4th, 2022"
output:
  BiocStyle::html_document:
    toc_float: true
package: epiregulon
vignette: >
  %\VignetteIndexEntry{Epiregulon}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
Gene regulatory networks model the underlying gene regulation hierarchies that drive gene expression and observed phenotypes. The main function of the epiregulon R package is to infer TF activity in single cells by constructing a gene regulatory network (regulons). This is achieved through integration of scATAC-seq and scRNA-seq data and incorporation of public bulk TF ChIP-seq data. Links between regulatory elements and their target genes are established by computing correlations between chromatin accessibility and gene expressions. 

Current prerequisite for running epiregulon is a ArchR project with pre-computed peak and gene matrices. It is also expected that LSI dimensionality reduction and integration with an scRNA-seq dataset has been performed. The scATAC-seq experiment can be either paired or unpaired with the scRNA-seq dataset as long as they were already integrated by ArchR. The final output of epiregulon is a matrix of TF activities where rows are individual TFs and columns are single cell indexes.

In this vignette we demonstrate the workflow of epiregulon along with some visualization functionalities using the [tutorial datasets](https://www.archrproject.com/articles/Articles/tutorial.html) from ArchR development team.

# Installation
```{r setup, message=FALSE}
devtools::install_github("OncoBioinfo-MO/Epiregulon", host = "https://github.roche.com/api/v3")
library(epiregulon)
```

# Data preparation


Please refer to the full ArchR [manual] (https://www.archrproject.com/bookdown/index.html) for instructions 

Before running Epiregulon, the following needs to be completed:
1. Obtain a peak matrix on scATACseq by using addGroupCoverages > addReproduciblePeakSet > addPeakMatrix (link)
2. RNAseq integration 
a. For multiome data, use addGeneExpressionMatrix 
b. For unpaired scATACseq, use addGeneIntegrationMatrix
3. Perform embedding from joint scRNAseq and scATACseq

To verify that all the necessary matrices are present,

library(ArchR)

archR_project_path="/gstore/project/lineage/sam/heme_GRN/OUTPUT"
proj<- loadArchRProject(path = outdir, showLogo = TRUE)
getAvailableMatrices(proj)

# Quick start

1. Retrieve bulk TF ChIP-seq binding sites 

First, we retrieve the information of TF binding sites collected from Cistrome and ENCODE ChIP-seq, which are hosted on Genomitory. Currently, human genomes HG19 and HG38 are available. We plan to add options for murine genomes and perhaps other model organisms in the near future. 
```{r}
grl <- getTFMotifInfo(genome="hg19")
head(grl)

```

2. Link ATACseq peaks to target genes

Next, we compute peak to gene correlations using the addPeak2GeneLinks function from ArchR package. The user would need to supply a path to an ArchR directory with a project contains peak and gene matrices, as well as already performed Latent semantic indexing (LSI) dimensionality reduction and integration with a scRNA-seq dataset. The example project shown here utilizes the [tutorial datasets](https://www.archrproject.com/articles/Articles/tutorial.html) provided by the ArchR development team. 
```{r}
# path to ArchR project
archR_project_path="/gstore/project/lineage/sam/heme_GRN/OUTPUT"
p2g <- getP2Glinks(archR_project_path)
head(p2g)
```

3. Add TF motif binding to peaks

The next step is to add the TF motif binding information retrieved previously to the ArchR object. This step calls the addPeakAnnotations function in ArchR to generate a binary matrix containing TF binding information for each peak to gene link. The file is stored as Annotations/ChIP-Matches-In-Peaks.rds in the ArchR project directory.
```{r}
addTFMotifInfo(archr_path = "/gstore/project/lineage/sam/heme_GRN/OUTPUT", grl)
```

4. Generate regulons

A long format dataframe, representing the inferred regulons, is then generated. The dataframe consists of three columns:

* tf (transcription factor)
* target gene
* peak to gene correlation between tf and target gene
```{r}
regulon <- getRegulon(archr_path = "/gstore/project/lineage/sam/heme_GRN/OUTPUT", p2g)
head(regulon)
```
