---
title: "Epiregulon"
author:
- name: Shang-Yang Chen
  email: chens179@gene.com
- name: Xiaosai Yao
  email: yaox19@gene.com
date: "January 17th, 2022"
output:
  BiocStyle::html_document:
    toc_float: true
package: epiregulon
vignette: >
  %\VignetteIndexEntry{Epiregulon}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
library(epiregulon)
```

# Introduction
Gene regulatory networks model the underlying gene regulation hierarchies that drive gene expression and observed phenotypes. The main function of the epiregulon R package is to infer TF activity in single cells by constructing a gene regulatory network (regulons). This is achieved through integration of scATAC-seq and scRNA-seq data and incorporation of public bulk TF ChIP-seq data. Links between regulatory elements and their target genes are established by computing correlations between chromatin accessibility and gene expressions. Current prerequisite for running epiregulon is a ArchR project with pre-computed peak and gene matrices. It is also expected that LSI dimensionality reduction and integration with an scRNA-seq dataset has been performed. The scATAC-seq experiment can be either paired or unpaired with the scRNA-seq dataset as long as they were already integrated by ArchR using addGeneExpressionMatrix (paired multi-ome) or addGeneIntegrationMatrix (unpaired). The final output of epiregulon is a matrix of TF activities where rows are individual TFs and columns are single cell indexes.

In this vignette we demonstrate the workflow of epiregulon along with some visualization functionalities using the [tutorial datasets](https://www.archrproject.com/articles/Articles/tutorial.html) from ArchR development team.

# Example
First, we retrieve the information of TF binding sites collected from Cistrome and ENCODE ChIP-seq, which are hosted on Genomitory. Currently, human genomes HG19 and HG38 are available. We plan to add options for murine genomes and perhaps other model organisms in the near future. 
```{r}
grl <- getTFMotifInfo(genome="hg19")
head(grl)

```

Next, we compute peak to gene correlations using the addPeak2GeneLinks function from ArchR package. The user would need to supply a path to an ArchR directory with a project contains peak and gene matrices, as well as already performed Latent semantic indexing (LSI) dimensionality reduction and integration with a scRNA-seq dataset. The example project shown here utilizes the [tutorial datasets](https://www.archrproject.com/articles/Articles/tutorial.html) provided by the ArchR development team. 
```{r}
# path to ArchR project
p2g <- getP2Glinks("/gstore/project/lineage/sam/heme_GRN/OUTPUT")
head(p2g)
```

The next step is to add the TF motif binding information retrieved previously to the ArchR object. This step calls the addPeakAnnotations function in ArchR to generate a binary matrix containing TF binding information for each peak to gene link. The file is stored as Annotations/ChIP-Matches-In-Peaks.rds in the ArchR project directory.
```{r}
addTFMotifInfo(archr_path = "/gstore/project/lineage/sam/heme_GRN/OUTPUT", grl)
```

A long format dataframe, representing the inferred regulons, is then generated. The dataframe consists of three columns:

* tf (transcription factor)
* target gene
* peak to gene correlation between tf and target gene
```{r}
regulon <- getRegulon(archr_path = "/gstore/project/lineage/sam/heme_GRN/OUTPUT", p2g)
head(regulon)
```

To infer activity of specific TFs using regulons just generated, we utilized a single cell CRISPRa Perturb-seq dataset from the manuscript [Exploring genetic interaction manifolds contructed from rich single-cell phenotypes](https://pubmed.ncbi.nlm.nih.gov/31395745/) as example. The regulons were trimmed to contain only TFs present in the Perturb-seq data. The user can supply any single cell gene expression dataset of their choice. 
```{r}
####### load filtered CRISPR data
sce=readRDS("/gstore/project/lineage/sam/heme_GRN/OUTPUT/small.sce.rds")

######## trim regulon
tf_unique=unique(c(sce$first_gRNA, sce$second_gRNA))
regulon=regulon %>% dplyr::filter(tf %in% tf_unique)
nrow(regulon)
```

Epiregulon outputs two different correlations. The first one was the correlations between chromatin accessibility of regulatory elements vs expression of target genes calculated by ArchR. The second one can be generated by the addWeights function, which compute the correlation between gene expressions of TF vs expressions of target genes. The user is required to supply the clustering or batch labels of the scRNA-seq dataset when running addWeights. 
```{r, warning=FALSE}
regulon.w=addWeights(regulon=regulon,
                     sce=sce,
                     cluster_factor="guide_identity",
                     block_factor=NULL,
                     corr=TRUE,
                     MI=FALSE,
                     multicore=TRUE)
head(regulon.w)
```

Finally, the activities for a specific TF in each cell are computed by averaging expressions of target genes linked to the TF weighted by the correlation variable of user's choice. 
$$y=\frac{1}{n}\sum_{i=1}^{n} x_i * corr_i$$
where $y$ is the activity of a TF for a cell
$n$ is the total number of targets for a TF
$x_i$ is the log count expression of target i where i in {1,2,...,n}
$corr_i$ is the weight of TF and target i

```{r}
score.combine <- calculateActivity(sce, regulon.w, "corr", method="weightedMean")
head(score.combine[,1:10])
```


# Session Info

```{r}
sessionInfo()
```


