---
title: "Multiome tutorial"
author:
- Shang-Yang Chen^[chens179@gene.com]
- Xiaosai Yao^[yaox19@gene.com]
date: "February 4th, 2022"
output: pdf_document
package: epiregulon
vignette: >
  %\VignetteIndexEntry{Multiome tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

# Introduction
This tutorial walks through an example of TF activity inference in single cell multiome data. This is a dataset generated by infecting LNCaP cells with NKX2-1 and GATA6 and examine the effects of these TFs on AR activity

# Installation

Epiregulon is currently available on R/dev

```{r, message=FALSE}
library(epiregulon)
```


```{r setup, message=FALSE, eval=FALSE}
devtools::install_gitlab(repo='scwg/gene-transcriptional-network/activity-inference/Epiregulon', 
                         auth_token = "<gitlab token>", 
                         host = "https://code.roche.com/" )

library(epiregulon)

```

# Data preparation


Please refer to the full ArchR [manual](https://www.archrproject.com/bookdown/index.html) for instructions 

Before running Epiregulon, the following analyses need to be completed:
1. Obtain a peak matrix on scATACseq by using addGroupCoverages > addReproduciblePeakSet > addPeakMatrix. See chapter [10](https://www.archrproject.com/bookdown/calling-peaks-with-archr.html) from ArchR manual
2. RNAseq integration. 
a. For unpaired scATACseq, use addGeneIntegrationMatrix. See chapter [8] (https://www.archrproject.com/bookdown/defining-cluster-identity-with-scrna-seq.html) from ArchR manual
b. For multiome data, use addGeneExpressionMatrix. See [multiome](https://greenleaflab.github.io/ArchR_2020/Ex-Analyze-Multiome.html) tutorial
3. Perform dimensionality reduction from with either single modalities or joint scRNAseq and scATACseq using [addCombinedDims](https://www.archrproject.com/reference/addCombinedDims.html)

Copy this ArchR project into your own directory
```{r}
archR_project_path="/gstore/project/ar_ligands/NE/reprogram_seq/multiome_arrayed/OUTPUT/doubletremoved/"
tutorial <- loadArchRProject(path = archR_project_path, showLogo = F)

# save tutorial data into your new directory and load it
myarchR_project_path = "/gstore/scratch/u/yaox19/multiome"
saveArchRProject(ArchRProj = tutorial, outputDirectory = myarchR_project_path, load = F )
proj <- loadArchRProject(path = myarchR_project_path, showLogo = F)

setwd(myarchR_project_path)
```

We verify that "GeneExpressionMatrix" and "PeakMatrix" are present for this tutorial. 

```{r}
getAvailableMatrices(proj)
```

We will use the joint reducedDims - "LSI_Combined" and  joint embeddings - "UMAP_Combined"
```{r}
proj@reducedDims
proj@embeddings

```

Retrieve gene expression matrix from the ArchR project
```{r}
GeneExpressionMatrix= getMatrixFromProject(
    ArchRProj = proj,
    useMatrix = "GeneExpressionMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)
```

Change to SingleCellExperiment
```{r}

GeneExpressionMatrix=as(GeneExpressionMatrix, "SingleCellExperiment")
assay(GeneExpressionMatrix, "logcounts") = assay(GeneExpressionMatrix, "GeneExpressionMatrix")

```
Transfer cell and gene information and embeddings from ArchR project to singleCellExperiment

```{r}
reducedDim(GeneExpressionMatrix, "UMAP_Combined")=getEmbedding(ArchRProj = proj, embedding = "UMAP_Combined", returnDF = TRUE)[colnames(GeneExpressionMatrix),]
colData(GeneExpressionMatrix)=getCellColData(proj)[colnames(GeneExpressionMatrix),]
rownames(GeneExpressionMatrix)=rowData(GeneExpressionMatrix)$name

```
Visualize singleCellExperiment by UMAP
```{r}
scater::plotReducedDim(GeneExpressionMatrix, dimred = "UMAP_Combined", text_by = "Clusters", colour_by = "Clusters")

```

# Quick start

1. Retrieve bulk TF ChIP-seq binding sites 

First, we retrieve the information of TF binding sites collected from Cistrome and ENCODE ChIP-seq, which are hosted on Genomitory. Currently, human genomes HG19 and HG38 are available. We plan to add options for murine genomes and perhaps other model organisms in the near future. 
```{r}
grl <- getTFMotifInfo(genome="hg38")
head(grl)

```

2. Link ATAC-seq peaks to target genes

Next, we compute peak to gene correlations using the addPeak2GeneLinks function from ArchR package. The user would need to supply a path to an ArchR directory with a project contains peak and gene matrices, as well as already performed Latent semantic indexing (LSI) dimensionality reduction and integration with a scRNA-seq dataset. 

```{r}
# path to ArchR project
p2g <- calculateP2G(ArchR_path = myarchR_project_path, useDim = "LSI_Combined", useMatrix = "GeneExpressionMatrix")
head(p2g)

```
Alternatively, users can now also supply peak, gene, and dimensional reduction matrices derived from a MultiAssayExperiment object to be compatible with future GPSA multiome workflow. Epiregulon implements a custom algorithm to derive a more stringent set of P2G correlations compared to ArchR. 

```{r, eval=FALSE}
# load the MAE object
mae <- readRDS("/gstore/project/archr_importer/ne_multiome/mae.rds")
# peak matrix
peakmatrix <- mae[["PeakMatrix"]]
# expression matrix
expmatrix <- mae[["GeneIntegrationMatrix"]]
rownames(expmatrix) <- rowData(expmatrix)$name
# dimensional reduction matrix
reducedDim <- SingleCellExperiment::reducedDims(mae[['TileMatrix500']])[["IterativeLSI"]]

p2g <- calculateP2G(peakmatrix, expmatrix, reducedDim)

head(p2g)
```
3. Add TF motif binding to peaks

The next step is to add the TF motif binding information by overlapping the regions of the peak matrix with the bulk chip-seq database loaded in 2. The user can supply either an archR project path and this function will retrieve the peak matrix, or a peakMatrix in the form of a Granges object or RangedSummarizedExperiment.

```{r}

overlap <- addTFMotifInfo(archR_project_path=myarchR_project_path, grl=grl, p2g=p2g)

```

4. Generate regulons

A long format dataframe, representing the inferred regulons, is then generated. The dataframe consists of three columns:

* tf (transcription factor)
* target gene
* peak to gene correlation between tf and target gene

```{r}

regulon <- getRegulon(p2g, overlap, aggregate=TRUE)
head(regulon)
```

Epiregulon outputs two different correlations. The first, termed "corr", is the correlation between chromatin accessibility of regulatory elements vs expression of target genes calculated by ArchR. The second, termed "weight", can be generated by the addWeights function, which compute the correlation between gene expressions of TF vs expressions of target genes, shown below. The user is required to supply the clustering or batch labels of the scRNA-seq dataset when running addWeights. "Weight" is the preferred metric for calculating activity.

```{r, warning=FALSE}


regulon.w <- addWeights(regulon=regulon,
                     sce=GeneExpressionMatrix,
                     cluster_factor="Clusters",
                     exprs_values="GeneExpressionMatrix",
                     block_factor=NULL,
                     corr=TRUE,
                     MI=FALSE,
                     BPPARAM=BiocParallel::MulticoreParam())
head(regulon.w)

```

5. Calculate TF activity 

Finally, the activities for a specific TF in each cell are computed by averaging expressions of target genes linked to the TF weighted by the correlation variable of user's choice. 
$$y=\frac{1}{n}\sum_{i=1}^{n} x_i * corr_i$$
where $y$ is the activity of a TF for a cell
$n$ is the total number of targets for a TF
$x_i$ is the log count expression of target i where i in {1,2,...,n}
$corr_i$ is the weight of TF and target i

```{r}
score.combine <- calculateActivity(GeneExpressionMatrix, regulon.w, "weight", method="weightedMean", assay="GeneExpressionMatrix")

```

6. Perform differential activity
```{r}

markers <- findDifferentialActivity(score.combine, GeneExpressionMatrix$Clusters, pval.type="some", direction="up", test.type= "t")
```

Take the top TFs
```{r}
markers.sig <- getSigGenes(markers, topgenes = 5 )

```
7. Visualize the results

First visualize the known differential TFs by bubble plot
```{r}
plotBubble(activity_matrix = score.combine, tf = c("NKX2-1","GATA6","FOXA1","FOXA2", "AR"), class=GeneExpressionMatrix$Clusters)
```

Then visualize the most differential TFs by clusters
```{r}
plotBubble(activity_matrix = score.combine, tf = markers.sig$tf, class=GeneExpressionMatrix$Clusters)

```

Visualize the known differential TFs by violin plot. Note there is no activity calculated for SOX2 because the expression of SOX2 is 0 in all cells.

```{r}
plotActivityViolin(score.combine, c("NKX2-1","GATA6","FOXA1","FOXA2", "AR", "SOX2"), GeneExpressionMatrix$Clusters)
```

Visualize the known differential TFs by UMAP
```{r, fig.height = 8, fig.width = 12}

plotActivityDim(sce=GeneExpressionMatrix, activity_matrix=score.combine, tf=  c("NKX2-1","GATA6","FOXA1","FOXA2", "AR", "SOX2"), dimtype="UMAP_Combined", label = "Clusters", point_size=0.1, ncol = 3)
```

8. Geneset enrichment
Sometimes we are interested to know what pathways are enriched in the regulon of a particular TF. We can perform geneset enrichment using the enricher function from [clusterProfiler](http://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html).

```{r, fig.height = 10, fig.width = 12}
#retrieve genesets
H = EnrichmentBrowser::getGenesets(org = "hsa", db = "msigdb", cat = "H", gene.id.type = "SYMBOL" )
C6 = EnrichmentBrowser::getGenesets(org = "hsa", db = "msigdb", cat = "C6", gene.id.type = "SYMBOL" )

#combine genesets and convert genesets to be compatible with enricher
gs=c(H,C6)
gs.list=do.call(rbind,lapply(names(gs), function(x) {data.frame(gs=x, genes=gs[[x]])}))

enrichresults=regulonEnrich(c("GATA6","AR"), regulon=regulon.w, corr="weight", corr_cutoff=0.5, genesets=gs.list)

#plot results
enrichPlot(results=enrichresults)
```

# Session Info

```{r}
sessionInfo()
```


